cwltool/task_queue.py

```diff
@@ -0,0 +1,119 @@
+# Copyright (C) The Arvados Authors. All rights reserved.
+#
+# SPDX-License-Identifier: Apache-2.0
+"""TaskQueue."""
+
+import queue
+import threading
+
+from typing import Callable, Optional
+
+from .loghandler import _logger
+
+
+class TaskQueue(object):
+    """
+    A TaskQueue class.
+
+    Uses a first-in, first-out queue of tasks executed on a fixed number of
+    threads.
+
+    If thread_count == 1 then as you add() tasks they will be immediately
+    executed. Otherwise new tasks enter the queue and will be executed in
+    the order received.
+
+    The thread_count is also used as the maximum size of the queue.
+
+    The threads are created during TaskQueue initialization, so delay the
+    creation of a TaskQueue until you really need it.
+
+
+    Attributes
+    ----------
+    in_flight
+        the number of tasks in the queue
+    """
+
+    def __init__(self, lock: threading.Lock, thread_count: int):
+        self.thread_count = thread_count
+        self.task_queue: queue.Queue[Optional[Callable[[], None]]] = queue.Queue(
+            maxsize=self.thread_count
+        )
+        self.task_queue_threads = []
+        self.lock = lock
+        self.in_flight = 0
+        self.error: Optional[Exception] = None
+
+        for _r in range(0, self.thread_count):
+            t = threading.Thread(target=self._task_queue_func)
+            self.task_queue_threads.append(t)
+            t.start()
+
+    def _task_queue_func(self) -> None:
+        while True:
+            task = self.task_queue.get()
+            if task is None:
+                return
+            try:
+                task()
+            except Exception as e:
+                _logger.exception("Unhandled exception running task")
+                self.error = e
+
+            with self.lock:
+                self.in_flight -= 1
+
+    def add(
+        self,
+        task: Callable[[], None],
+        unlock: Optional[threading.Lock] = None,
+        check_done: Optional[threading.Event] = None,
+    ) -> None:
+        """
+        Add your task to the queue.
+
+        If the TaskQueue was created with only one thread then your task will
+        be immediately executed.
+
+        The optional unlock will be released prior to attempting to add the
+        task to the queue.
+
+        If the optional "check_done" threading.Event's flag is set, then we
+        will skip adding this task to the queue.
+        """
+        if self.thread_count > 1:
+            with self.lock:
+                self.in_flight += 1
+        else:
+            task()
+            return
+
+        while True:
+            try:
+                if unlock is not None:
+                    unlock.release()
+                if check_done is not None:
+                    if check_done.is_set():
+                        return
+                self.task_queue.put(task, block=True, timeout=3)
```

Comment:
I'm surprised that you can't put more tasks into the queue than `thread_count` was set to.

What's the reasoning?

Reply:
If you have N threads that are not doing anything yet, you can put up to N things in the queue and then each thread will have something to do.

If you have more than N things, you are going to end up blocking eventually because all the threads are busy.  This just means you have N things running and up to another N things queued before you block.  Making the queue unbounded size is unlikely to have any advantage in throughput, but it would mean you could end up using a bit more RAM if you created a  backlog of hundreds of pending jobs.